vignettes/BERT_basics.Rmd

```diff
@@ -0,0 +1,165 @@
+
+## What is BERT?
+
+For now, think of BERT as a function that takes in text and puts out numbers--a
+long list of numbers for *each* token (~word) of the input.
+
+In fact, there is really a family of such functions. Google research released
+the first BERT models in late 2018, and others have followed (as well as many
+that include slight modifications to the original structure). When referring to
+any results using "BERT", it is important to specify which BERT you're talking
+about.
+
+## The input to BERT
+
+### Tokenization
+
+BERT takes "natural" text as input, with some restrictions. The first thing that
+BERT does is to
+[tokenize](https://en.wikipedia.org/wiki/Lexical_analysis#Tokenization) the
+input text. Most common words will be tokenized as themselves, but words that
+are not included in the vocabulary of that particular version of BERT will be
+tokenized as two or more "word pieces". Each character of punctuation also gets
+its own token. Finally, BERT adds a few special delimiter tokens to each piece
+of input.
+
+For example, "Who doesn't like tacos?" might be tokenized as:
+
+`[CLS] who doesn ' t like ta ##cos ? [SEP]`
+
+The "##" bit indicates that "cos" was originally attached to "ta". The word
+"tacos" was split up this way because that word isn't found in (this version of)
+BERT's limited vocabulary.
+
+Any output from BERT is organized in terms of tokens like this.
+
+### Sequences
+
+Current BERT models can process chunks of text that consist of no more than 512
+tokens (so the maximum number of words is rather fewer than that in practice). 
+If you have text longer than that, you will need to find some way of splitting
+it up. A natural way of splitting up your text is by individual sentences, if
+possible. In this package (consistently) and in other literature (perhaps less
+consistently) a chunk of text processed by BERT is referred to as a "sequence".
+So a list of such chunks may be indexed by a "sequence_index", for example.
+
+### Segments
+
+A sequence *may* be divided into two [^segments] "segments". This is useful when
+your particular application calls for two distinct pieces of text to be input
+(e.g. a model that evaluates the logical compatibility of two statements). Note
+that the sequence as a whole still can't exceed the token limit, so splitting
+your text into segments is not a way to input longer text. In fact, a delimiter
+token is required to separate the two segments, which counts against the 512
+limit, so you'll actually lose a bit of capacity by using segments.
+
+When possible, it's probably best to use individual sentences as your input
+sequences (or segments, if you're going that way). BERT was trained at the
+sentence level, and you're less likely to hit the token limit with individual
+sentences than with, say, paragraphs of text.
+
+[^segments]: In principle, the input could have any number of segments, but the
+original BERT models were limited to two segments.
```

Comment:
```suggestion
BERT models are limited to two segments.
```
I think make this firmer until/unless we implement something that accepts more than 2.